File: Problems/142_Linked_List_Cycle_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *slow = head;
        ListNode *fast = head;
        while(fast && fast->next){
            slow = slow->next;
            fast = fast->next->next;
            if(slow==fast){
                fast = head;
                if(slow==fast) return slow;
                while(fast!=slow){
                    fast = fast->next;
                    slow = slow->next;
                    if(slow==fast) return slow;
                }
            }
        }
        return NULL;
    }
};

File: Problems/142_Linked_List_Cycle_II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        fast = head
        slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if fast==slow:
                slow2 = head
                if slow == slow2:
                    return slow
                while slow!=slow2:
                    slow = slow.next
                    slow2 = slow2.next
                return slow
        return None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        seen = []
        while head:
            if head in seen:
                return head
            seen.append(head)
            head = head.next